pkg/servicemesh/federation/controller/controller.go

```diff
@@ -0,0 +1,284 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package controller
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	xnsinformers "github.com/maistra/xns-informer/pkg/informers"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry"
+	"istio.io/istio/pilot/pkg/serviceregistry/aggregate"
+	"istio.io/istio/pilot/pkg/serviceregistry/federation"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/kube"
+	kubecontroller "istio.io/istio/pkg/kube/controller"
+	"istio.io/istio/pkg/servicemesh/apis/servicemesh/v1alpha1"
+	clientsetservicemeshv1alpha1 "istio.io/istio/pkg/servicemesh/client/v1alpha1/clientset/versioned"
+	informersservicemeshv1alpha1 "istio.io/istio/pkg/servicemesh/client/v1alpha1/informers/externalversions/servicemesh/v1alpha1"
+	"istio.io/pkg/log"
+)
+
+func init() {
+	schemasBuilder := collection.NewSchemasBuilder()
+	schemasBuilder.MustAdd(collections.IstioNetworkingV1Alpha3Destinationrules)
+	schemasBuilder.MustAdd(collections.IstioNetworkingV1Alpha3Virtualservices)
+	schemasBuilder.MustAdd(collections.IstioNetworkingV1Alpha3Gateways)
+	schemasBuilder.MustAdd(collections.IstioSecurityV1Beta1Authorizationpolicies)
+	// XXX: we should consider adding this directly to the service registry
```

Comment:
+1 we might want to create two registry entries directly for proxied services. should we file an issue for it?

Reply:
Yeah.  I think it complicates things because the federation controller then also becomes a service registry itself, so it will need to integrate as both a config store and service registry.

Reply:
I've created an issue for this, MAISTRA-2298